nemo/collections/tts/models/fastpitch_e2e.py

```diff
@@ -0,0 +1,873 @@
+    def convert_text_to_waveform(
+        self, tokens: 'torch.tensor', speaker: Optional[int] = None, pace: float = 1.0
+    ) -> torch.tensor:
+        if self.training:
+            logging.warning("convert_text_to_waveform() is meant to be called in eval mode.")
+        if isinstance(speaker, int):
+            speaker = torch.tensor([speaker]).to(self.device)
+        audio = self(text=tokens, durs=None, pitch=None, speaker=speaker, pace=pace)
+        return audio
+
+    def training_step(self, batch, batch_idx):
+        attn_prior, durs, speaker, energy = None, None, None, None
+        if self.learn_alignment:
+            assert self.ds_class_name == "TTSDataset", f"Unknown dataset class: {self.ds_class_name}"
+
+            batch_dict = process_batch(batch, self._train_dl.dataset.sup_data_types_set)
+            # print(batch_dict)
+            audio = batch_dict.get("audio")
+            audio_lens = batch_dict.get("audio_lens")
+            text = batch_dict.get("text")
+            text_lens = batch_dict.get("text_lens")
+            attn_prior = batch_dict.get("align_prior_matrix", None)
+            pitch = batch_dict.get("pitch", None)
+            energy = batch_dict.get("energy", None)
+            speaker = batch_dict.get("speaker_id", None)
+            durs = batch_dict.get("durations", None)
+        else:
+            audio, audio_lens, text, text_lens, attn_prior, pitch, durs, speaker = batch
+
+        mels, spec_len = self.fp_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        (
+            mels_pred,
+            dec_lens,
+            durs_predicted,
+            log_durs_pred,
+            pitch_pred,
+            attn_soft,
+            attn_logprob,
+            attn_hard,
+            attn_hard_dur,
+            pitch,
+            energy_pred,
+            energy_tgt,
+        ) = self.fp_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            energy=energy,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=spec_len,
+            input_lens=text_lens,
+        )
+        if durs is None:
+            durs = attn_hard_dur
+
+        audio = audio.unsqueeze(1)
+
+        fp_inter_mel = self.mel_loss_fn(spect_predicted=mels_pred, spect_tgt=mels)
+
+        # Skip batch if it is too short
+        if self.cfg.use_patches:
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
+            if len(ids_str_max[ids_str_max < 0]) > 0:
+                print('Encountered short batch')
+                return
+
+            mels_pred, ids_slice = rand_slice_segments(
+                mels_pred, dec_lens, self._cfg.segment_size // self._cfg.n_window_stride
+            )
+            audio = slice_segments(audio, ids_slice * self.cfg.n_window_stride, self._cfg.segment_size)
```

Comment:
Nitpick: it will be easier to read if values like `self._cfg.segment_size // self._cfg.n_window_stride` and `ids_slice * self.cfg.n_window_stride` are saved into variables with appropriate names.

Reply:
This code is adopted from VITS implementation, do u think we need to change it in both places then?